app/openvpn_client_lwt.ml

```diff
@@ -88,17 +113,49 @@ let read_from_fd fd =
       let cs = Cstruct.of_bytes ~len:count buf in
       Logs.debug (fun m -> m "read %d bytes" count) ;
       Lwt.return cs)
-  |> Lwt_result.map_err (fun e ->
-      Rresult.R.msgf "read error %s" (Printexc.to_string e))
+  |> Lwt_result.map_err (fun e -> `Msg (Printexc.to_string e))
 
-let rec reader mvar fd =
+let rec reader_tcp mvar fd =
   read_from_fd fd >>= function
   | Error `Msg msg ->
     Logs.err (fun m -> m "read error from remote %s" msg);
     Lwt_mvar.put mvar `Connection_failed
   | Ok data ->
     Lwt_mvar.put mvar (`Data data) >>= fun () ->
-    reader mvar fd
+    reader_tcp mvar fd
+
+let read_udp =
+  let bufsize = 65535 in
+  let buf = Bytes.create bufsize in
+  fun (sa, fd) ->
+    Lwt_result.catch (
+      Lwt_unix.recvfrom fd buf 0 bufsize [] >>= fun (count, sa') ->
+      if count = 0 then
+        Lwt.fail_with "end of file from server"
```

Comment:
```suggestion
        Lwt.fail_with "UDP zero-length payload from server"
```
For UDP this is a payload of length 0 rather than a connection shutdown (like it would be for TCP), but if receive 0-length packets then something is probably off, so unless we see a lot of these (keep-alive?) I think it makes sense to keep this exception case.

The only problem I can think of is that this is not guarded by `sa = sa'`, so 0-length payloads from any peer will presently kill the connection, I think maybe that should be changed.